packetbeat/protos/dns/dns.go

```diff
@@ -707,3 +721,170 @@ func decodeDnsPacket(data []byte) (dns *layers.DNS, err error) {
+	data, err := decodeDnsData(stream.data)
+
+	if err != nil {
+		logp.Debug("dns", NonDnsCompleteMsg+" addresses %s, length %d",
+			tcpTuple.String(), len(stream.data))
+
+		// wait for decoding with the next segment
+		return priv
+	}
+	dns.messageComplete(tcpTuple, dir, stream, data)
+
+	return priv
+}
+
+func (dns *Dns) messageComplete(tcpTuple *common.TcpTuple, dir uint8, s *DnsStream, decodedData *layers.DNS) {
+	dns.handleDns(s.message, tcpTuple, dir, s.data, decodedData)
+
+	s.PrepareForNewMessage()
+}
+
+func (dns *Dns) handleDns(m *DnsMessage, tcpTuple *common.TcpTuple, dir uint8, data []byte, decodedData *layers.DNS) {
+	dnsTuple := DnsTupleFromIpPort(&m.Tuple, TransportTcp, decodedData.ID)
+	m.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(tcpTuple.IpPort())
+	m.Data = decodedData
+	m.Length = len(data)
+
+	if decodedData.QR == Query {
+		dns.receivedDnsRequest(&dnsTuple, m)
+	} else /* Response */ {
+		dns.receivedDnsResponse(&dnsTuple, m)
+	}
+}
+
+func (stream *DnsStream) PrepareForNewMessage() {
+	stream.message = nil
+}
+
+func (dns *Dns) ReceivedFin(tcpTuple *common.TcpTuple, dir uint8, private protos.ProtocolData) protos.ProtocolData {
+	if private == nil {
+		return private
+	}
+	dnsData, ok := private.(dnsPrivateData)
+	if !ok {
+		return private
+	}
+	if dnsData.Data[dir] == nil {
+		return dnsData
+	}
+	stream := dnsData.Data[dir]
+	if stream.message != nil {
+		decodedData, err := decodeDnsData(stream.data)
+		if err == nil {
+			dns.messageComplete(tcpTuple, dir, stream, decodedData)
+		} else /*Failed decode */ {
+			if dir == tcp.TcpDirectionReverse {
+				dns.publishDecodeFailureNotes(dnsData)
+				stream.PrepareForNewMessage()
+			}
+			logp.Debug("dns", NonDnsCompleteMsg+" addresses %s, length %d",
+				tcpTuple.String(), len(stream.data))
+		}
+	}
+
+	return dnsData
+}
+
+func (dns *Dns) GapInStream(tcpTuple *common.TcpTuple, dir uint8, nbytes int, private protos.ProtocolData) (priv protos.ProtocolData, drop bool) {
+	dnsData, ok := private.(dnsPrivateData)
+
+	if !ok {
+		return private, false
+	}
+
+	stream := dnsData.Data[dir]
+
+	if stream == nil || stream.message == nil {
+		return private, false
+	}
+
+	decodedData, err := decodeDnsData(stream.data)
+
+	// Add Notes if the failed stream is the response
+	if err != nil {
+		if dir == tcp.TcpDirectionReverse {
+			dns.publishDecodeFailureNotes(dnsData)
+		}
+
+		// drop the stream because it is binary and it would be rare to have a decodable message later
+		logp.Debug("dns", NonDnsCompleteMsg+" addresses %s, length %d",
+			tcpTuple.String(), len(stream.data))
+		return private, true
+	}
+
+	// publish and ignore the gap. No case should reach this code though ...
+	dns.messageComplete(tcpTuple, dir, stream, decodedData)
+	return private, false
+}
+
+// Add Notes to the query stream about a failure to decode the response
+func (dns *Dns) publishDecodeFailureNotes(dnsData dnsPrivateData) {
+	streamOrigin := dnsData.Data[tcp.TcpDirectionOriginal]
```

Comment:
```streamReverse``` is checked before the method is called (in Gap and Fin). It will be checked a second time in ```publishDecodeFailureNotes```, except if you don't think it's proper.